src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,379 @@
+                final List<Consumer<String>> toNotify;
+                synchronized (consumerListeners) {
+                    toNotify = consumerListeners.containsKey(unlocked) ?
+                            new ArrayList<>(consumerListeners.get(unlocked)) : null;
+                }
+                if (null != toNotify) {
+                    for (final Consumer<String> listener : toNotify) {
+                        try {
+                            listener.accept(unlocked);
+                        } catch (final RuntimeException ex) {
+                            LOG.error("Failed to notify about event type {} unlock", unlocked, ex);
+                        }
+                    }
+                }
+            }
+            try {
+                zooKeeperHolder.get().setData().forPath(
+                        toZkPath("/nodes/" + thisId), String.valueOf(change.version).getBytes(Charsets.UTF_8));
+            } catch (final Exception ex) {
+                LOG.error("Failed to update node version in zk. Will try to reprocess again", ex);
+                return;
+            }
+            queuedChanges.poll();
+            LOG.info("Delayed change {} successfully processed", change);
+        }
+    }
+
+    private void checkForNewChange() {
+        if (newVersionPresent.compareAndSet(true, false)) {
+            runLocked(() -> {
+                boolean success = false;
+                try {
+                    queuedChanges.add(
+                            new DelayedChange(readData("/version", this::versionChanged, Integer::parseInt),
+                                    this.zooKeeperHolder.get().getChildren().forPath(toZkPath("/locked_et"))));
+                    success = true;
+                } catch (final RuntimeException ex) {
+                    throw ex;
+                } catch (final InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(ex);
+                } catch (final Exception ex) {
+                    throw new RuntimeException(ex);
+                } finally {
+                    if (!success) {
+                        newVersionPresent.set(true);
+                    }
+                }
+            });
+        }
+    }
+
+    private void versionChanged(final WatchedEvent ignore) {
+        LOG.info("Adding refresh call to delayed changes list");
+        newVersionPresent.set(true);
+    }
+
+    private void runLocked(final Runnable action) {
+        try {
+            Exception releaseException = null;
+            final InterProcessLock lock =
+                    new InterProcessSemaphoreMutex(zooKeeperHolder.get(), ROOT_PATH + "/lock");
+            lock.acquire();
+            try {
+                action.run();
+            } finally {
+                try {
+                    lock.release();
+                } catch (final Exception ex) {
+                    LOG.warn("Failed to release lock", ex);
```

Comment:
Shouldn't this be give a higher severity? Presumably, if we can't release the lock, it's still there, and processes waiting for it will be stuck.